nano/node/messages.cpp

```diff
@@ -1581,4 +1604,151 @@ std::size_t nano::node_id_handshake::size (nano::message_header const & header_a
+ */
+
+nano::asc_pull_req::asc_pull_req (const nano::network_constants & constants) :
+	message (constants, nano::message_type::asc_pull_req)
+{
+}
+
+nano::asc_pull_req::asc_pull_req (bool & error, nano::stream & stream, const nano::message_header & header) :
+	message (header)
+{
+	error = deserialize (stream);
+}
+
+void nano::asc_pull_req::visit (nano::message_visitor & visitor) const
+{
+	visitor.asc_pull_req (*this);
+}
+
+void nano::asc_pull_req::serialize (nano::stream & stream) const
+{
+	header.serialize (stream);
+	nano::write (stream, type);
+	nano::write (stream, id);
+	nano::write (stream, start);
+}
+
+bool nano::asc_pull_req::deserialize (nano::stream & stream)
+{
+	debug_assert (header.type == nano::message_type::asc_pull_req);
+	bool error = false;
+	try
+	{
+		nano::read (stream, type);
+		nano::read (stream, id);
+		nano::read (stream, start);
+	}
+	catch (std::runtime_error const &)
+	{
+		error = true;
+	}
+	return error;
+}
+
+/*
+ * asc_pull_ack
+ */
+
+nano::asc_pull_ack::asc_pull_ack (const nano::network_constants & constants) :
+	message (constants, nano::message_type::asc_pull_ack)
+{
+	update_header ();
+}
+
+nano::asc_pull_ack::asc_pull_ack (bool & error, nano::stream & stream, const nano::message_header & header) :
+	message (header)
+{
+	error = deserialize (stream);
+	if (!error)
+	{
+		update_header ();
+	}
+}
+
+void nano::asc_pull_ack::visit (nano::message_visitor & visitor) const
+{
+	visitor.asc_pull_ack (*this);
+}
+
+void nano::asc_pull_ack::serialize (nano::stream & stream) const
+{
+	debug_assert (header.extensions.to_ulong () > 0); // Block payload must have least `not_a_block` terminator
+	header.serialize (stream);
+	nano::write (stream, type);
+	nano::write (stream, id);
+	serialize_blocks (stream);
+}
+
+void nano::asc_pull_ack::serialize_blocks (nano::stream & stream) const
+{
+	debug_assert (blocks_m.size () <= max_blocks);
+	for (auto & block : blocks_m)
+	{
+		debug_assert (block != nullptr);
+		nano::serialize_block (stream, *block);
+	}
+	// For convenience, end with null block terminator
+	nano::serialize_block_type (stream, nano::block_type::not_a_block);
+}
+
+std::vector<std::shared_ptr<nano::block>> nano::asc_pull_ack::blocks () const
+{
+	return blocks_m;
+}
+
+void nano::asc_pull_ack::update_header ()
+{
+	std::vector<uint8_t> bytes;
+	{
+		nano::vectorstream blocks_stream (bytes);
+		serialize_blocks (blocks_stream);
+	}
+	debug_assert (bytes.size () <= 65535u); // Max int16 for storing size
+	debug_assert (bytes.size () >= 1); // At least `not_a_block` terminator
+	header.extensions = std::bitset<16> (bytes.size ());
+}
+
+void nano::asc_pull_ack::blocks (std::vector<std::shared_ptr<nano::block>> & blocks)
+{
+	blocks_m = blocks;
+	update_header ();
+}
+
+bool nano::asc_pull_ack::deserialize (nano::stream & stream)
+{
+	debug_assert (header.type == nano::message_type::asc_pull_ack);
+	bool error = false;
+	try
+	{
+		nano::read (stream, type);
+		nano::read (stream, id);
```

Comment:
We should be specifying the endianness of the ID and not just assuming it will be little endian.

Reply:
In the latest version I ensured that the serialization always uses little endian, as it's the most common endianness out there